### ISSUE TITLE
Need to be more careful about deleting unreachable blocks

### ISSUE THREAD

**Issue by AndyAyersMS** [dotnet/llilc#489]
In `ReaderBase::fgRemoveUnusedBlocks` we remove unreachable blocks and associated LLVM IR. This needs to be done more carefully than the way it is done now.

Test JIT\jit64\gc\fgtest2.exe has the following method failure:
```
Failed to read T.Main[fgDeleteEdge]
```
because there are two unreachable blocks and one branches to the other. This creates an ordering dependence which we currently violate. Of course in general unreachable code may contain cycles making it impossible to choose any particular order.

Seems like we should call LLVM's `removeUnreachableBlocks` function here (which does a multi-pass removal) instead of what we're doing now.

**Comment by Bevinsky** [comment]
Forgive me if I've missed some detail, but isn't the reason for the failure because the `fgDeleteEdge` method simply isn't implemented in GenIR? See https://github.com/dotnet/llilc/blob/master/include/Reader/readerir.h#L585.

I agree that using `removeUnreachableBlocks` is a better choice. Perhaps it would be prudent to remove the block elimination in ReaderBase and let the client define how it wants to eliminate dead blocks by overriding `fgRemoveUnusedBlocks` instead?

If no one else is interested I might try my hand at it.

**Comment by AndyAyersMS** [comment]
I thought to implement `fgDeleteEdge` initially too, but then decided the pattern in ReaderBase was not general enough to be workable with LLVM. 

@Bevinsky my only hesitation at leaving this entirely this up to the client layer is that the base layer has been tracking which blocks are visited and `fgRemoveUnusedBlocks` uses this visited data to remove blocks, and then clears the visited data. 

So if we let the client layer do its own determination on what blocks to remove we might want to double-check somehow that the base layer agrees, and then ensure the either that visited bits on the blocks that remain are cleared, or that nothing else depends on these bits later on.

Feel free to take this one . I don't think we can directly assign it to you (yet) but just leave a note here if you want to work on it.

**Comment by Bevinsky** [comment]
As I see it, the benefit of using `removeUnreachableBlocks` is twofold, in that it is a tried and tested part of LLVM and is less prone to cause issues, and also that it seems to insert some unreachability instructions for dead code removal as well.

Why would the visited data matter after dead block elimination if we clear it anyway? 

I would like to try using `removeUnreachableBlocks`, but AFAIK there isn't a way to get what blocks were removed in an independent manner. It could be done in a client method that returns what blocks were removed, and then the base can compare this to its own result.

**Comment by roterdam** [comment]
I may be mistaken, but isn't this just calling `removeUnreachableBlocks(*Function)` instead of `fgRemoveUnusedBlocks` ? You can also just call simplyCFG, but of course that may not be desirable in debug code.

**Comment by Bevinsky** [comment]
@roterdam, I think the idea is to have client-independent block removal, with the actual removal of individual blocks and edges implemented in the client. So, at the very least `fgRemoveUnusedBlocks` still has to be called by `msilToIR`, and then it needs to be overridden in the IR client which calls `removeUnreachableBlocks`.

**Comment by AndyAyersMS** [comment]
Right, we're trying to keep ReaderBase be independent of any particular IR, with the idea that down the road somebody could use ReaderBase and a different IR for some other purpose (for instance, a verifier might use a simpler IR and simpler type system). So we want ReaderBase to specify what behavior it needs and let the client figure out how to make it happen.

And I agree using `removeUnreachableBlocks` is the right thing to do here -- just trying to make sure it does more or less the same thing we're doing already.

**Comment by roterdam** [comment]
Right. I agree, and tried to implement it:

Hitting this assert.

https://github.com/Microsoft/llvm/blob/master/include/llvm/IR/Instructions.h#L2328

 ` Value *removeIncomingValue(const BasicBlock *BB, bool DeletePHIIfEmpty=true) {`
`    int Idx = getBasicBlockIndex(BB);`
    `assert(Idx >= 0 && "Invalid basic block argument to remove!");`

**Comment by Bevinsky** [comment]
@roterdam, Yeah, I tried the same thing and got the same assertion.

It seems that the IR generation is making some assumptions that the unvisitable basic blocks (according to the IL) will be removed, so this produces code like

    ; <label>:22                                      ; preds = %20
      .... Some code
      ret void
    
      br label %24  ; unreachable!
    
    ; <label>:24                                      ; preds = %11, %12, %13, %23
      %126 = phi i64 [ %2, %11 ], [ %51, %12 ], [ %3, %13 ]

(This isn't exactly what was in my output, I didn't save it. I wrote this from memory.)

Note that the phi instruction doesn't contain any value for the block 23. `removeUnreachableBlocks` chokes on this via `removePredecessor` and `getBasicBlockIndex`.

I'm not entirely familiar with the codebase, so I'm not sure where these phi instructions are being produced. Perhaps it's worth taking a look at to avoid making this assumption.
